internals/plan/plan.go

```diff
@@ -1021,82 +1022,111 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
 
 var fnameExp = regexp.MustCompile("^([0-9]{3})-([a-z](?:-?[a-z0-9]){2,}).yaml$")
 
-func ReadLayersDir(dirname string) ([]*Layer, error) {
+type LayerReader struct {
+	labels map[string]string
+	offset int
+}
+
+func (lr *LayerReader) ReadDir(dirname string) ([]*Layer, error) {
+	if lr.labels == nil {
+		lr.labels = make(map[string]string)
+	}
+
 	finfos, err := ioutil.ReadDir(dirname)
 	if err != nil {
 		// Errors from package os generally include the path.
 		return nil, fmt.Errorf("cannot read layers directory: %v", err)
 	}
 
 	orders := make(map[int]string)
-	labels := make(map[string]int)
+	labels := make(map[string]string)
+	// Copy already loaded layer labels with full path to the layer file
+	// as they have come from a different import path.
+	for k, v := range lr.labels {
+		labels[k] = v
+	}
 
 	// Documentation says ReadDir result is already sorted by name.
 	// This is fundamental here so if reading changes make sure the
 	// sorting is preserved.
 	var layers []*Layer
 	for _, finfo := range finfos {
-		if finfo.IsDir() || !strings.HasSuffix(finfo.Name(), ".yaml") {
+		fileName := finfo.Name()
+		if finfo.IsDir() || !strings.HasSuffix(fileName, ".yaml") {
 			continue
 		}
 		// TODO Consider enforcing permissions and ownership here to
 		//      avoid mistakes that could lead to hacks.
-		match := fnameExp.FindStringSubmatch(finfo.Name())
+		match := fnameExp.FindStringSubmatch(fileName)
 		if match == nil {
-			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", finfo.Name())
+			return nil, fmt.Errorf("invalid layer filename: %q (must look like \"123-some-label.yaml\")", fileName)
 		}
 
-		data, err := ioutil.ReadFile(filepath.Join(dirname, finfo.Name()))
+		filePath := filepath.Join(dirname, fileName)
+		data, err := ioutil.ReadFile(filePath)
```

Comment:
Could use the newer `os.ReadFile` while we're in here (now that we're on Go 1.20; ioutil was deprecated in 1.16).